network/src/protocols/wire/proxy_protocol.rs

```diff
@@ -0,0 +1,369 @@
+    let address_size = address_size.as_ref().read_u16::<BigEndian>()?;
+
+    // To prevent excessive reads if the address size is to large, drop it
+    // This is not in the spec, but allows for all types in the spec
+    if address_size > UNIX_SIZE {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            "Address size too large",
+        ));
+    }
+
+    let mut address_bytes: Vec<u8> = vec![0; address_size as usize];
+    stream.read_exact(&mut address_bytes).await?;
+    let mut address_bytes = address_bytes.as_slice();
+
+    let source_address = match family_and_protocol {
+        // TODO: Support UDP in the future
+        LOCAL_PROTOCOL | UDP_IPV4 | UDP_IPV6 | TCP_UNIX | UDP_UNIX => {
+            // UNSPEC, UDP, and UNIX Steam/datagram
+            // Accept connection but ignore address info as per spec
+            original_addr.clone()
+        }
+        TCP_IPV4 => {
+            // This is not mentioned in the spec, but if it doesn't match we might not read correctly
+            if address_size != IPV4_SIZE {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    "Address size doesn't match type",
+                ));
+            }
+
+            let src_addr = address_bytes.read_u32::<BigEndian>()?;
+            let _dest_addr = address_bytes.read_u32::<BigEndian>()?;
+            let src_port = address_bytes.read_u16::<BigEndian>()?;
+            let _dest_port = address_bytes.read_u16::<BigEndian>()?;
+            let src_addr = Ipv4Addr::from(src_addr);
+
+            // TODO: What do we want to do about the destination information?
+            let socket_addr = SocketAddr::new(IpAddr::V4(src_addr), src_port);
+            NetworkAddress::from(socket_addr)
+        }
+        TCP_IPV6 => {
+            // This is not mentioned in the spec, but if it doesn't match we might not read correctly
+            if address_size != IPV6_SIZE {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    "Address size doesn't match type",
+                ));
+            }
+
+            let src_addr = address_bytes.read_u128::<BigEndian>()?;
+            let _dest_addr = address_bytes.read_u128::<BigEndian>()?;
+            let src_port = address_bytes.read_u16::<BigEndian>()?;
+            let _dest_port = address_bytes.read_u16::<BigEndian>()?;
+            let src_addr = Ipv6Addr::from(src_addr);
+
+            // TODO: What do we want to do about the destination information?
+            let socket_addr = SocketAddr::new(IpAddr::V6(src_addr), src_port);
+            NetworkAddress::from(socket_addr)
+        }
+        _ => {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Unsupported Address Family or Protocol",
+            ));
+        }
+    };
+
+    Ok(source_address)
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
do you know if proxy protocol lists any test vectors we could also sanity check against?

Reply:
It does not, but it does list this code in C++ at the bottom of the spec, it's just kinda messy.

I read through the whole spec and basically enumerated the cases we want to cover in these tests.